### PR TITLE
freetype: add mirror url + cache cmake + del fPIC if shared + partially fix cmake_find_package_multi

### PR DESCRIPTION
**Pull request by SpaceIm**
Specify library name and version:  **freetype/all**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by SpaceIm** [comment]
This recipe has a TODO to enable harbuzz dependency, but harbuzz also depends on freetype. How to solve this cyclic dependency in conan?
By the way, Freetype can also optionally depends on brotli.

**Comment by danimtb** [comment]
> This recipe has a TODO to enable harfbuzz dependency, but harfbuzz also depends on freetype. How to solve this cyclic dependency in conan?

I think this was solved before using and option to optionally depend on harfbuzz, but I don't know how that would work when the option is enabled 🤔

**Comment by SSE4** [comment]
this is somehow needed to break a vicious circle, e.g.:
- build `harfbuzz` (without freetype) -> build `freetype` (with harfbuzz) -> build `harfbuzz` (with freetype) -> discard first `harfbuzz` build
- build `freetype` (without harfbuzz) -> build `harfbuzz` (with freetype) -> build `freetype` (with harfbuzz) -> discard first `freetype` build

**Comment by SpaceIm** [comment]
@SSE4 How could it be implemented in CCI recipes? Seems hard to achieve in conan without breaking current dependency graph resolution.

**Comment by SSE4** [comment]
@SpaceIm to be honest, I don't know if it's even possible right now, or if it requires additional feature(s) from conan's side (or maybe conan-center-index special handling, or hook?). I defer this question to @danimtb and @jgsogo :)

File: recipes/freetype/all/conanfile.py [hunk 1]
```diff
@@ -122,6 +134,6 @@ def package_info(self):
         self.env_info.FT2_CONFIG = freetype_config
         self.user_info.LIBTOOL_VERSION = self._libtool_version
         self._chmod_plus_x(freetype_config)
-        self.cpp_info.names['cmake_find_package'] = 'Freetype'
-        self.cpp_info.names['cmake_find_package_multi'] = 'Freetype'
-        self.cpp_info.names['pkg_config'] = 'freetype2'
+        self.cpp_info.names["cmake_find_package"] = "Freetype"
+        self.cpp_info.names["cmake_find_package_multi"] = "freetype" # TODO: CMake config file has a non namespaced imported target
```
Comment (jgsogo) [inline recipes/freetype/all/conanfile.py:138]: ¿????? Targets generated by `cmake_find_package` and `cmake_find_package_multi` should be the same... what's happening here? Is it a bug in Conan client?

**Comment by jgsogo** [comment]
freetype - harfbuzz, how do they work together?

Are they linking to each other? Is it something like...?
 * freetype generates libraries `freetype_core` and `freetype`
 * harfbuzz generates libraries `harfbuzz_core` and `harfbuzz`
 * `freetype` links with `harfbuzz_core`
 * `harfbuzz` links with `freetype_core`

File: recipes/freetype/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/freetype/all/conanfile.py:138]: Official module file is `FindFreetype.cmake`, while official config file is `freetype-config.cmake`, therefore `cmake_find_package` and `cmake_find_package_multi` shouldn't be the same.

Comment (jgsogo) [reply recipes/freetype/all/conanfile.py:138]: OMG! .... so 
 * `find_package(Freetype)` using `MODULE` mode, but
 * `find_package(freetype NO_MODULE)` or `find_package(freetype CONFIG)`

---

Right now I'm ~~sad 😭~~ confused. I'm not sure if we could ever merge these two generators into one...

Comment (uilianries) [reply recipes/freetype/all/conanfile.py:138]: When you think Protobuf is the only weird project based on cmake modules ...

**Comment by SSE4** [comment]
```
Installed Libraries:
libharfbuzz.so, libharfbuzz-gobject.so, libharfbuzz-icu.so, and libharfbuzz-subset.so
```
```
Installed Library:
libfreetype.so
```

**Comment by SSE4** [comment]
after doing LFS recommended steps:
```
sse4@sse4-VirtualBox:~/ft-build/lib$ ldd libfreetype.so
        linux-vdso.so.1 (0x00007fff01d36000)
        libharfbuzz.so.0 => /home/sse4/ft-build/lib/libharfbuzz.so.0 (0x00007f030079f000)
        libpng16.so.16 => /usr/lib/x86_64-linux-gnu/libpng16.so.16 (0x00007f030056d000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f0300350000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f02fff5f000)
        libglib-2.0.so.0 => /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0 (0x00007f02ffc48000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f02ff8aa000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f0300d5b000)
        libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007f02ff638000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f02ff419000)
```
```
sse4@sse4-VirtualBox:~/ft-build/lib$ ldd libharfbuzz.so
        linux-vdso.so.1 (0x00007ffc90957000)
        libglib-2.0.so.0 => /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0 (0x00007f0366df3000)
        libfreetype.so.6 => /home/sse4/ft-build/lib/libfreetype.so.6 (0x00007f0366b45000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f0366754000)
        libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007f03664e2000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f03662c3000)
        libpng16.so.16 => /usr/lib/x86_64-linux-gnu/libpng16.so.16 (0x00007f0366091000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f0365e74000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f0367418000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f0365ad6000)
```
what exactly does it mean?
1. yes, they link to each other (libharfbuzz.so -> libfreetype.so, libfreetype.so -> libharfbuzz.so)
2. yes, there is a real circular dependency
3. no, there are no "core" libraries and "plug-ins". there are just two libraries, one for freetype, one for harfbuzz

**Comment by SSE4** [comment]
however, this is still getting deep into details of harfbuzz vs freetype. to be honest, that's not something special or uncommon. there are many examples like that. one additional example is OpenCV + ffmpeg : you may build ffmpeg with OpenCV support (to get some image processing filters), and you may build OpenCV with ffmpeg support (to allow to import video files). and you still need to break cycle, so you do:
- ffmpeg (temporary) -> opencv -> ffmpeg
- opencv (temporary) -> ffmpeg -> opencv

@jgsogo still, it doesn't answer how should I model that (typical) scenario with conan?
does it require additional development in conan itself, conan center index or conan hook?
if so, how would it look like from the user's perspective (e.g. which conan commands should I run to get the desired effect)?

**Comment by SSE4** [comment]
digging deeper into the use-case:
first, apply simple patch to get harfbuzz option into the freetype recipe:
```
diff --git a/recipes/freetype/all/conanfile.py b/recipes/freetype/all/conanfile.py
index d16c093..762a1b0 100644
--- a/recipes/freetype/all/conanfile.py
+++ b/recipes/freetype/all/conanfile.py
@@ -21,13 +21,15 @@ class FreetypeConan(ConanFile):
         "with_png": [True, False],
         "with_zlib": [True, False],
         "with_bzip2": [True, False],
+        "with_harfbuzz": [True, False],
     }
     default_options = {
         'shared': False,
         'fPIC': True,
         'with_png': True,
         'with_zlib': True,
-        'with_bzip2': True
+        'with_bzip2': True,
+        'with_harfbuzz': False
     }
     _source_subfolder = "source_subfolder"
     _build_subfolder = "build_subfolder"
@@ -39,6 +41,8 @@ class FreetypeConan(ConanFile):
             self.requires("zlib/1.2.11")
         if self.options.with_bzip2:
             self.requires("bzip2/1.0.8")
+        if self.options.with_harfbuzz:
+            self.requires('harfbuzz/2.6.8')

     def config_options(self):
         if self.settings.os == "Windows":
@@ -62,8 +66,8 @@ class FreetypeConan(ConanFile):
         cmake.definitions["FT_WITH_BZIP2"] = self.options.with_bzip2
         cmake.definitions["CMAKE_DISABLE_FIND_PACKAGE_BZip2"] = not self.options.with_bzip2
         # TODO: Harfbuzz can be added as an option as soon as it is available.
-        cmake.definitions["FT_WITH_HARFBUZZ"] = False
-        cmake.definitions["CMAKE_DISABLE_FIND_PACKAGE_HarfBuzz"] = True
+        cmake.definitions["FT_WITH_HARFBUZZ"] = self.options.with_harfbuzz
+        cmake.definitions["CMAKE_DISABLE_FIND_PACKAGE_HarfBuzz"] = not self.options.with_harfbuzz
         cmake.configure(build_dir=self._build_subfolder)
         return cmake
```
run the following commands:

- `conan create freetype/all freetype/2.10.2@ -o freetype:with_harfbuzz=False -k`
- `conan create harfbuzz/all harfbuzz/2.6.8@ -o harfbuzz:with_freetype=True -k`
- `conan create freetype/all freetype/2.10.2@ -o freetype:with_harfbuzz=True -k`

the last one fails:
```
ERROR: Loop detected in context host: 'harfbuzz/2.6.8' requires 'freetype/2.10.2' which is an ancestor too
```

how to break this loop in conan?
the only idea I have right now is to build first freetype with a different name (e.g. `freetype-temporary`).
this obviously might render in the following conan graph:
```
freetype -> harfbuzz -> freetype-temporary
```
now I need to remove `freetype-temporary` (okay, that's easy part, just do `conan remove freetype-temporary/* -f`).

but still, there is a reference to the `freetype-temporary` from the `harfbuzz` which I need to **override**.

in that step, how am I supposed to hack into graph to change that reference? I assume conan's override won't work, as it can only override versions (I am not sure).

@jgsogo WDYT?

**Comment by jgsogo** [comment]
Thanks a lot for the investigation, @SSE4 . I really don't know how to fix this loop with the current Conan paradigm (if not using a temporary recipe, which is something really ugly from the CCI POV). How to build **interdependent packages?**

I think this will take us some time, meanwhile we can merge these PRs and some user will complain if they try to build this option combination. But we cannot stop here.

**Pull request merged**